panic occurs when nested object declared in schema

### Module version

```
➜ terraform -version
Terraform v1.0.6
on darwin_amd64
+ provider hashicorp.com/edu/hashicups v0.3.2
```

### Relevant provider source code

```go
// Order Resource schema
func (r resourceOrderType) GetSchema(_ context.Context) (schema.Schema, []*tfprotov6.Diagnostic) {
	return schema.Schema{
		Attributes: map[string]schema.Attribute{
			"id": {
				Type: types.StringType,
				// When Computed is true, the provider will set value --
				// the user cannot define the value
				Computed: true,
			},
			"last_updated": {
				Type:     types.StringType,
				Computed: true,
			},
			"items": {
				// If Required is true, Terraform will throw error if user
				// doesn't specify value
				// If Optional is true, user can choose to supply a value
				Required: true,
				Attributes: schema.ListNestedAttributes(map[string]schema.Attribute{
					"quantity": {
						Type:     types.NumberType,
						Required: true,
					},
					"coffee": {
						Required: true,
						Attributes: schema.SingleNestedAttributes(map[string]schema.Attribute{
							"id": {
								Type:     types.NumberType,
								Required: true,
							},
							"name": {
								Type:     types.StringType,
								Computed: true,
							},
							"teaser": {
								Type:     types.StringType,
								Computed: true,
							},
							"description": {
								Type:     types.StringType,
								Computed: true,
							},
							"price": {
								Type:     types.NumberType,
								Computed: true,
							},
							"image": {
								Type:     types.StringType,
								Computed: true,
							},
							"ingredients": {
								Optional: true,
								Attributes: schema.ListNestedAttributes(map[string]schema.Attribute{
									"ingredient_id": {
										Type:     types.NumberType,
										Optional: true,
									},
									"name": {
										Type:     types.StringType,
										Optional: true,
									},
									"quantity": {
										Type:     types.NumberType,
										Optional: true,
									},
									"unit": {
										Type:     types.StringType,
										Optional: true,
									},
								}, schema.ListNestedAttributesOptions{}),
							},
						}),
					},
				}, schema.ListNestedAttributesOptions{}),
			},
		},
	}, nil
}
```

### Terraform Configuration Files
<!--
Paste the relevant parts of your Terraform configuration between the ``` marks below.

For large Terraform configs, please use a service like Dropbox and share a link to the ZIP file. For security, you can also encrypt the files using our GPG public key.
-->

```hcl
terraform {
  required_providers {
    hashicups = {
      version = "~> 0.3.1"
      source  = "hashicorp.com/edu/hashicups"
    }
  }
}
provider "hashicups" {
  username = "education"
  password = "test123"
  host     = "http://localhost:19090"
}

resource "hashicups_order" "edu" {
  items = [{
    coffee = {
      id = 3

      ingredients = [{
        name = "sugar"
      }]

    }
    quantity = 2
    }, {
    coffee = {
      id = 1
    }
    quantity = 2
    }
  ]
}

output "edu_order" {
  value = hashicups_order.edu
}

```

### Debug Output

[Crash Log on Gist](https://gist.github.com/yufeiminds/67973778e3c0f700f960013eaed3dc1e#file-crash-951271563-log)

### Expected Behavior

No crash for nested objects.

### Actual Behavior

```
hashicups_order.edu: Refreshing state... [id=1]
panic: inconsistent list element types (cty.Object(map[string]cty.Type{"coffee":cty.Object(map[string]cty.Type{"description":cty.String, "id":cty.Number, "image":cty.String, "ingredients":cty.List(cty.Object(map[string]cty.Type{"ingredient_id":cty.Number, "name":cty.String, "quantity":cty.Number, "unit":cty.String})), "name":cty.String, "price":cty.Number, "teaser":cty.String}), "quantity":cty.Number}) then cty.Object(map[string]cty.Type{"coffee":cty.Object(map[string]cty.Type{"description":cty.String, "id":cty.Number, "image":cty.String, "ingredients":cty.List(cty.ObjectWithOptionalAttrs(map[string]cty.Type{"ingredient_id":cty.Number, "name":cty.String, "quantity":cty.Number, "unit":cty.String}, []string{"ingredient_id", "name", "quantity", "unit"})), "name":cty.String, "price":cty.Number, "teaser":cty.String}), "quantity":cty.Number}))

goroutine 131 [running]:
github.com/zclconf/go-cty/cty.ListVal(0xc0003569c0, 0x2, 0x2, 0x2f88240, 0xc00039cfc0, 0x37babf8, 0xc000353940)
	/Users/distiller/go/pkg/mod/github.com/zclconf/go-cty@v1.9.1/cty/value_init.go:166 +0x57e
github.com/hashicorp/terraform/internal/plans/objchange.proposedNewNestedType(0xc0004704e0, 0x37bab88, 0xc000353c00, 0x2e66040, 0xc00000e2b8, 0x37bab88, 0xc000353a00, 0x2e66040, 0xc00000e288, 0xc00000e288, ...)
	/Users/distiller/project/project/internal/plans/objchange/objchange.go:346 +0x6d3
github.com/hashicorp/terraform/internal/plans/objchange.proposedNewAttributes(0xc000907380, 0x37babf8, 0xc000353c10, 0x2f88240, 0xc00039cc00, 0x37babf8, 0xc000353a30, 0x2f88240, 0xc00039c990, 0x0)
	/Users/distiller/project/project/internal/plans/objchange/objchange.go:295 +0x49a
github.com/hashicorp/terraform/internal/plans/objchange.proposedNew(0xc0009073e0, 0x37babf8, 0xc000353c10, 0x2f88240, 0xc00039cc00, 0x37babf8, 0xc000353a30, 0x2f88240, 0xc00039c990, 0x20, ...)
	/Users/distiller/project/project/internal/plans/objchange/objchange.go:80 +0x165
github.com/hashicorp/terraform/internal/plans/objchange.ProposedNew(0xc0009073e0, 0x37babf8, 0xc000353c10, 0x2f88240, 0xc00039cc00, 0x37babf8, 0xc000353a30, 0x2f88240, 0xc00039c990, 0x0, ...)
	/Users/distiller/project/project/internal/plans/objchange/objchange.go:42 +0x11f
github.com/hashicorp/terraform/internal/terraform.(*NodeAbstractResourceInstance).plan(0xc000737340, 0x37ec5a0, 0xc0001d2b60, 0x0, 0xc0003dfa40, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/distiller/project/project/internal/terraform/node_resource_abstract_instance.go:732 +0xdab
github.com/hashicorp/terraform/internal/terraform.(*NodePlannableResourceInstance).managedResourceExecute(0xc0002f90e0, 0x37ec5a0, 0xc0001d2b60, 0x100c408, 0x32fd020, 0xc000a1bc50)
	/Users/distiller/project/project/internal/terraform/node_resource_plan_instance.go:196 +0x8cb
github.com/hashicorp/terraform/internal/terraform.(*NodePlannableResourceInstance).Execute(0xc0002f90e0, 0x37ec5a0, 0xc0001d2b60, 0xd6ddc02, 0xc000a1bce8, 0x100d305, 0x2fd3580)
	/Users/distiller/project/project/internal/terraform/node_resource_plan_instance.go:54 +0xb3
github.com/hashicorp/terraform/internal/terraform.(*ContextGraphWalker).Execute(0xc0006588f0, 0x37ec5a0, 0xc0001d2b60, 0xd6ddc30, 0xc0002f90e0, 0x0, 0x0, 0x0)
	/Users/distiller/project/project/internal/terraform/graph_walk_context.go:129 +0xbf
github.com/hashicorp/terraform/internal/terraform.(*Graph).walk.func1(0x32fd020, 0xc0002f90e0, 0x0, 0x0, 0x0)
	/Users/distiller/project/project/internal/terraform/graph.go:59 +0xbd3
github.com/hashicorp/terraform/internal/dag.(*Walker).walkVertex(0xc0003df200, 0x32fd020, 0xc0002f90e0, 0xc0006d32c0)
	/Users/distiller/project/project/internal/dag/walk.go:381 +0x288
created by github.com/hashicorp/terraform/internal/dag.(*Walker).Update
	/Users/distiller/project/project/internal/dag/walk.go:304 +0x1246




!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
A crash log has been placed at "crash.951271563.log" relative to your current
working directory. It would be immensely helpful if you could please
report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version. That
information is available on the first line of crash.log. You can also
get it by running 'terraform --version' on the command line.

SECURITY WARNING: the "crash.951271563.log" file that was created may contain
sensitive information that must be redacted before it is safe to share
on the issue tracker.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

### Steps to Reproduce
<!--
Please list the full steps required to reproduce the issue, for example:
1. `terraform init`
2. `terraform apply`
-->

Pre-requirements: Start the hashicups server according to []

1. Clone my fork: [yufeiminds/terraform-provider-hashicups-pf](https://github.com/yufeiminds/terraform-provider-hashicups-pf)
2. `make install`
3. `cd examples && terraform init`
4. `terraform apply`
5. See the panic that occurs

### References

- https://github.com/hashicorp/terraform/issues/27275
- https://github.com/hashicorp/terraform/issues/27269
- https://github.com/hashicorp/terraform/issues/26195
- https://github.com/hashicorp/terraform/issues/28171

Thanks for added details @yufeiminds!

That still leaves me with a different stack trace than shown above, triggered during the decoding of the provider response:

```
panic: inconsistent list element types (cty.Object(map[string]cty.Type{"coffee":cty.Object(map[string]cty.Type{"description":cty.String, "id":cty.Number, "image":cty.String, "ingredients":cty.List(cty.Object(map[string]cty.Type{"ingredient_id":cty.Number, "name":cty.String, "quantity":cty.Number, "unit":cty.String})), "name":cty.String, "price":cty.Number, "teaser":cty.String}), "quantity":cty.Number}) then cty.Object(map[string]cty.Type{"coffee":cty.Object(map[string]cty.Type{"description":cty.String, "id":cty.Number, "image":cty.String, "ingredients":cty.List(cty.ObjectWithOptionalAttrs(map[string]cty.Type{"ingredient_id":cty.Number, "name":cty.String, "quantity":cty.Number, "unit":cty.String}, []string{"ingredient_id", "name", "quantity", "unit"})), "name":cty.String, "price":cty.Number, "teaser":cty.String}), "quantity":cty.Number}))

goroutine 118 [running]:
github.com/zclconf/go-cty/cty.ListVal({0x14000737780, 0x2, 0x2})
	/Users/jbardin/go/pkg/mod/github.com/zclconf/go-cty@v1.9.1/cty/value_init.go:166 +0x4e4
github.com/zclconf/go-cty/cty/msgpack.unmarshalList(0x14000796768, {{0x102a4bf70, 0x140009f8fc0}}, {0x14000a83210, 0x1, 0x1})
	/Users/jbardin/go/pkg/mod/github.com/zclconf/go-cty@v1.9.1/cty/msgpack/unmarshal.go:161 +0x4e8
github.com/zclconf/go-cty/cty/msgpack.unmarshal(0x14000796768, {{0x102a4bf00, 0x140009f8fd0}}, {0x14000a83210, 0x1, 0x1})
	/Users/jbardin/go/pkg/mod/github.com/zclconf/go-cty@v1.9.1/cty/msgpack/unmarshal.go:52 +0x464
github.com/zclconf/go-cty/cty/msgpack.unmarshalObject(0x14000796768, 0x14000b566f0, {0x0, 0x0, 0x0})
	/Users/jbardin/go/pkg/mod/github.com/zclconf/go-cty@v1.9.1/cty/msgpack/unmarshal.go:297 +0x500
github.com/zclconf/go-cty/cty/msgpack.unmarshal(0x14000796768, {{0x102a4bf70, 0x14000a83200}}, {0x0, 0x0, 0x0})
	/Users/jbardin/go/pkg/mod/github.com/zclconf/go-cty@v1.9.1/cty/msgpack/unmarshal.go:60 +0x6b4
github.com/zclconf/go-cty/cty/msgpack.Unmarshal({0x14000398200, 0xf7, 0x100}, {{0x102a4bf70, 0x14000a83200}})
	/Users/jbardin/go/pkg/mod/github.com/zclconf/go-cty@v1.9.1/cty/msgpack/unmarshal.go:22 +0xc8
github.com/hashicorp/terraform/internal/plugin6.decodeDynamicValue(0x14000123bc0, {{0x102a4bf70, 0x14000a83200}})
	/Users/jbardin/terraform/internal/plugin6/grpc_provider.go:612 +0xcc
github.com/hashicorp/terraform/internal/plugin6.(*GRPCProvider).PlanResourceChange(0x14000899a80, {{0x140004e9b80, 0xf}, {{{0x102a4bf70, 0x140009f8ce0}}, {0x0, 0x0}}, {{{0x102a4bf70, 0x140009f9830}}, {0x10266f620, ...}}, ...})
	/Users/jbardin/terraform/internal/plugin6/grpc_provider.go:426 +0x724
github.com/hashicorp/terraform/internal/terraform.(*NodeAbstractResourceInstance).plan(0x140009f2000, {0x102aadf20, 0x140002902a0}, 0x0, 0x0, 0x0, {0x0, 0x0, 0x0})
	/Users/jbardin/terraform/internal/terraform/node_resource_abstract_instance.go:755 +0xdac
...
```

The panic message however is the similar, and will have a related root cause. It seems that the response from the provider is not properly encoded, however that should only result in an error, not a panic, so we need to fix that in the cty package too.

I would really like to get a reproduction of the original stack trace shown above too, so if you have any more details there it would be a great help. Now that the class of error is evident, we may be able to uncover where the error lies in Terraform by manually auditing the code too.

Thanks!

--- a/internal/backend/unparsed_value_test.go
+++ b/internal/backend/unparsed_value_test.go
@@ -24,9 +24,10 @@ func TestParseVariableValuesUndeclared(t *testing.T) {
 	}
 	decls := map[string]*configs.Variable{
 		"declared1": {
-			Name:        "declared1",
-			Type:        cty.String,
-			ParsingMode: configs.VariableParseLiteral,
+			Name:           "declared1",
+			Type:           cty.String,
+			ConstraintType: cty.String,
+			ParsingMode:    configs.VariableParseLiteral,
 			DeclRange: hcl.Range{
 				Filename: "fake.tf",
 				Start:    hcl.Pos{Line: 2, Column: 1, Byte: 0},
@@ -34,9 +35,10 @@ func TestParseVariableValuesUndeclared(t *testing.T) {
 			},
 		},
 		"missing1": {
-			Name:        "missing1",
-			Type:        cty.String,
-			ParsingMode: configs.VariableParseLiteral,
+			Name:           "missing1",
+			Type:           cty.String,
+			ConstraintType: cty.String,
+			ParsingMode:    configs.VariableParseLiteral,
 			DeclRange: hcl.Range{
 				Filename: "fake.tf",
 				Start:    hcl.Pos{Line: 3, Column: 1, Byte: 0},
@@ -44,10 +46,11 @@ func TestParseVariableValuesUndeclared(t *testing.T) {
 			},
 		},
 		"missing2": {
-			Name:        "missing1",
-			Type:        cty.String,
-			ParsingMode: configs.VariableParseLiteral,
-			Default:     cty.StringVal("default for missing2"),
+			Name:           "missing1",
+			Type:           cty.String,
+			ConstraintType: cty.String,
+			ParsingMode:    configs.VariableParseLiteral,
+			Default:        cty.StringVal("default for missing2"),
 			DeclRange: hcl.Range{
 				Filename: "fake.tf",
 				Start:    hcl.Pos{Line: 4, Column: 1, Byte: 0},

--- a/internal/configs/configschema/decoder_spec.go
+++ b/internal/configs/configschema/decoder_spec.go
@@ -121,7 +121,7 @@ func (b *Block) DecoderSpec() hcldec.Spec {
 			// implied type more complete, but if there are any
 			// dynamically-typed attributes inside we must use a tuple
 			// instead, at the expense of our type then not being predictable.
-			if blockS.Block.ImpliedType().HasDynamicTypes() {
+			if blockS.Block.specType().HasDynamicTypes() {
 				ret[name] = &hcldec.BlockTupleSpec{
 					TypeName: name,
 					Nested:   childSpec,
@@ -155,7 +155,7 @@ func (b *Block) DecoderSpec() hcldec.Spec {
 			// implied type more complete, but if there are any
 			// dynamically-typed attributes inside we must use a tuple
 			// instead, at the expense of our type then not being predictable.
-			if blockS.Block.ImpliedType().HasDynamicTypes() {
+			if blockS.Block.specType().HasDynamicTypes() {
 				ret[name] = &hcldec.BlockObjectSpec{
 					TypeName:   name,
 					Nested:     childSpec,
@@ -195,7 +195,7 @@ func (a *Attribute) decoderSpec(name string) hcldec.Spec {
 			panic("Invalid attribute schema: NestedType and Type cannot both be set. This is a bug in the provider.")
 		}
 
-		ty := a.NestedType.ImpliedType()
+		ty := a.NestedType.specType()
 		ret.Type = ty
 		ret.Required = a.Required || a.NestedType.MinItems > 0
 		return ret

--- a/internal/configs/configschema/implied_type.go
+++ b/internal/configs/configschema/implied_type.go
@@ -8,11 +8,23 @@ import (
 // ImpliedType returns the cty.Type that would result from decoding a
 // configuration block using the receiving block schema.
 //
+// The type returned from Block.ImpliedType differs from the type returned by
+// hcldec.ImpliedType in that there will be no objects with optional
+// attributes, since this value is not to be used for the decoding of
+// configuration.
+//
 // ImpliedType always returns a result, even if the given schema is
 // inconsistent. Code that creates configschema.Block objects should be
 // tested using the InternalValidate method to detect any inconsistencies
 // that would cause this method to fall back on defaults and assumptions.
 func (b *Block) ImpliedType() cty.Type {
+	return b.specType().WithoutOptionalAttributesDeep()
+}
+
+// specType returns the cty.Type used for decoding a configuration
+// block using the receiving block schema. This is the type used internally by
+// hcldec to decode configuration.
+func (b *Block) specType() cty.Type {
 	if b == nil {
 		return cty.EmptyObject
 	}
@@ -41,14 +53,20 @@ func (b *Block) ContainsSensitive() bool {
 	return false
 }
 
-// ImpliedType returns the cty.Type that would result from decoding a NestedType
-// Attribute using the receiving block schema.
+// ImpliedType returns the cty.Type that would result from decoding a
+// NestedType Attribute using the receiving block schema.
 //
 // ImpliedType always returns a result, even if the given schema is
 // inconsistent. Code that creates configschema.Object objects should be tested
 // using the InternalValidate method to detect any inconsistencies that would
 // cause this method to fall back on defaults and assumptions.
 func (o *Object) ImpliedType() cty.Type {
+	return o.specType().WithoutOptionalAttributesDeep()
+}
+
+// specType returns the cty.Type used for decoding a NestedType Attribute using
+// the receiving block schema.
+func (o *Object) specType() cty.Type {
 	if o == nil {
 		return cty.EmptyObject
 	}
@@ -56,7 +74,7 @@ func (o *Object) ImpliedType() cty.Type {
 	attrTys := make(map[string]cty.Type, len(o.Attributes))
 	for name, attrS := range o.Attributes {
 		if attrS.NestedType != nil {
-			attrTys[name] = attrS.NestedType.ImpliedType()
+			attrTys[name] = attrS.NestedType.specType()
 		} else {
 			attrTys[name] = attrS.Type
 		}

--- a/internal/configs/configschema/implied_type_test.go
+++ b/internal/configs/configschema/implied_type_test.go
@@ -112,6 +112,36 @@ func TestBlockImpliedType(t *testing.T) {
 				}),
 			}),
 		},
+		"nested objects with optional attrs": {
+			&Block{
+				Attributes: map[string]*Attribute{
+					"map": {
+						Optional: true,
+						NestedType: &Object{
+							Nesting: NestingMap,
+							Attributes: map[string]*Attribute{
+								"optional":          {Type: cty.String, Optional: true},
+								"required":          {Type: cty.Number, Required: true},
+								"computed":          {Type: cty.List(cty.Bool), Computed: true},
+								"optional_computed": {Type: cty.Map(cty.Bool), Optional: true, Computed: true},
+							},
+						},
+					},
+				},
+			},
+			// The ImpliedType from the type-level block should not contain any
+			// optional attributes.
+			cty.Object(map[string]cty.Type{
+				"map": cty.Map(cty.Object(
+					map[string]cty.Type{
+						"optional":          cty.String,
+						"required":          cty.Number,
+						"computed":          cty.List(cty.Bool),
+						"optional_computed": cty.Map(cty.Bool),
+					},
+				)),
+			}),
+		},
 	}
 
 	for name, test := range tests {
@@ -137,6 +167,211 @@ func TestObjectImpliedType(t *testing.T) {
 			&Object{},
 			cty.EmptyObject,
 		},
+		"attributes": {
+			&Object{
+				Nesting: NestingSingle,
+				Attributes: map[string]*Attribute{
+					"optional":          {Type: cty.String, Optional: true},
+					"required":          {Type: cty.Number, Required: true},
+					"computed":          {Type: cty.List(cty.Bool), Computed: true},
+					"optional_computed": {Type: cty.Map(cty.Bool), Optional: true, Computed: true},
+				},
+			},
+			cty.Object(
+				map[string]cty.Type{
+					"optional":          cty.String,
+					"required":          cty.Number,
+					"computed":          cty.List(cty.Bool),
+					"optional_computed": cty.Map(cty.Bool),
+				},
+			),
+		},
+		"nested attributes": {
+			&Object{
+				Nesting: NestingSingle,
+				Attributes: map[string]*Attribute{
+					"nested_type": {
+						NestedType: &Object{
+							Nesting: NestingSingle,
+							Attributes: map[string]*Attribute{
+								"optional":          {Type: cty.String, Optional: true},
+								"required":          {Type: cty.Number, Required: true},
+								"computed":          {Type: cty.List(cty.Bool), Computed: true},
+								"optional_computed": {Type: cty.Map(cty.Bool), Optional: true, Computed: true},
+							},
+						},
+						Optional: true,
+					},
+				},
+			},
+			cty.Object(map[string]cty.Type{
+				"nested_type": cty.Object(map[string]cty.Type{
+					"optional":          cty.String,
+					"required":          cty.Number,
+					"computed":          cty.List(cty.Bool),
+					"optional_computed": cty.Map(cty.Bool),
+				}),
+			}),
+		},
+		"nested object-type attributes": {
+			&Object{
+				Nesting: NestingSingle,
+				Attributes: map[string]*Attribute{
+					"nested_type": {
+						NestedType: &Object{
+							Nesting: NestingSingle,
+							Attributes: map[string]*Attribute{
+								"optional":          {Type: cty.String, Optional: true},
+								"required":          {Type: cty.Number, Required: true},
+								"computed":          {Type: cty.List(cty.Bool), Computed: true},
+								"optional_computed": {Type: cty.Map(cty.Bool), Optional: true, Computed: true},
+								"object": {
+									Type: cty.ObjectWithOptionalAttrs(map[string]cty.Type{
+										"optional": cty.String,
+										"required": cty.Number,
+									}, []string{"optional"}),
+								},
+							},
+						},
+						Optional: true,
+					},
+				},
+			},
+			cty.Object(map[string]cty.Type{
+				"nested_type": cty.Object(map[string]cty.Type{
+					"optional":          cty.String,
+					"required":          cty.Number,
+					"computed":          cty.List(cty.Bool),
+					"optional_computed": cty.Map(cty.Bool),
+					"object":            cty.Object(map[string]cty.Type{"optional": cty.String, "required": cty.Number}),
+				}),
+			}),
+		},
+		"NestingList": {
+			&Object{
+				Nesting: NestingList,
+				Attributes: map[string]*Attribute{
+					"foo": {Type: cty.String, Optional: true},
+				},
+			},
+			cty.List(cty.Object(map[string]cty.Type{"foo": cty.String})),
+		},
+		"NestingMap": {
+			&Object{
+				Nesting: NestingMap,
+				Attributes: map[string]*Attribute{
+					"foo": {Type: cty.String},
+				},
+			},
+			cty.Map(cty.Object(map[string]cty.Type{"foo": cty.String})),
+		},
+		"NestingSet": {
+			&Object{
+				Nesting: NestingSet,
+				Attributes: map[string]*Attribute{
+					"foo": {Type: cty.String},
+				},
+			},
+			cty.Set(cty.Object(map[string]cty.Type{"foo": cty.String})),
+		},
+		"deeply nested NestingList": {
+			&Object{
+				Nesting: NestingList,
+				Attributes: map[string]*Attribute{
+					"foo": {
+						NestedType: &Object{
+							Nesting: NestingList,
+							Attributes: map[string]*Attribute{
+								"bar": {Type: cty.String},
+							},
+						},
+					},
+				},
+			},
+			cty.List(cty.Object(map[string]cty.Type{"foo": cty.List(cty.Object(map[string]cty.Type{"bar": cty.String}))})),
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			got := test.Schema.ImpliedType()
+			if !got.Equals(test.Want) {
+				t.Errorf("wrong result\ngot:  %#v\nwant: %#v", got, test.Want)
+			}
+		})
+	}
+}
+
+func TestObjectContainsSensitive(t *testing.T) {
+	tests := map[string]struct {
+		Schema *Object
+		Want   bool
+	}{
+		"object contains sensitive": {
+			&Object{
+				Attributes: map[string]*Attribute{
+					"sensitive": {Sensitive: true},
+				},
+			},
+			true,
+		},
+		"no sensitive attrs": {
+			&Object{
+				Attributes: map[string]*Attribute{
+					"insensitive": {},
+				},
+			},
+			false,
+		},
+		"nested object contains sensitive": {
+			&Object{
+				Attributes: map[string]*Attribute{
+					"nested": {
+						NestedType: &Object{
+							Nesting: NestingSingle,
+							Attributes: map[string]*Attribute{
+								"sensitive": {Sensitive: true},
+							},
+						},
+					},
+				},
+			},
+			true,
+		},
+		"nested obj, no sensitive attrs": {
+			&Object{
+				Attributes: map[string]*Attribute{
+					"nested": {
+						NestedType: &Object{
+							Nesting: NestingSingle,
+							Attributes: map[string]*Attribute{
+								"public": {},
+							},
+						},
+					},
+				},
+			},
+			false,
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			got := test.Schema.ContainsSensitive()
+			if got != test.Want {
+				t.Errorf("wrong result\ngot:  %#v\nwant: %#v", got, test.Want)
+			}
+		})
+	}
+
+}
+
+// Nested attribute should return optional object attributes for decoding.
+func TestObjectSpecType(t *testing.T) {
+	tests := map[string]struct {
+		Schema *Object
+		Want   cty.Type
+	}{
 		"attributes": {
 			&Object{
 				Nesting: NestingSingle,
@@ -265,74 +500,10 @@ func TestObjectImpliedType(t *testing.T) {
 
 	for name, test := range tests {
 		t.Run(name, func(t *testing.T) {
-			got := test.Schema.ImpliedType()
+			got := test.Schema.specType()
 			if !got.Equals(test.Want) {
 				t.Errorf("wrong result\ngot:  %#v\nwant: %#v", got, test.Want)
 			}
 		})
 	}
-}
-
-func TestObjectContainsSensitive(t *testing.T) {
-	tests := map[string]struct {
-		Schema *Object
-		Want   bool
-	}{
-		"object contains sensitive": {
-			&Object{
-				Attributes: map[string]*Attribute{
-					"sensitive": {Sensitive: true},
-				},
-			},
-			true,
-		},
-		"no sensitive attrs": {
-			&Object{
-				Attributes: map[string]*Attribute{
-					"insensitive": {},
-				},
-			},
-			false,
-		},
-		"nested object contains sensitive": {
-			&Object{
-				Attributes: map[string]*Attribute{
-					"nested": {
-						NestedType: &Object{
-							Nesting: NestingSingle,
-							Attributes: map[string]*Attribute{
-								"sensitive": {Sensitive: true},
-							},
-						},
-					},
-				},
-			},
-			true,
-		},
-		"nested obj, no sensitive attrs": {
-			&Object{
-				Attributes: map[string]*Attribute{
-					"nested": {
-						NestedType: &Object{
-							Nesting: NestingSingle,
-							Attributes: map[string]*Attribute{
-								"public": {},
-							},
-						},
-					},
-				},
-			},
-			false,
-		},
-	}
-
-	for name, test := range tests {
-		t.Run(name, func(t *testing.T) {
-			got := test.Schema.ContainsSensitive()
-			if got != test.Want {
-				t.Errorf("wrong result\ngot:  %#v\nwant: %#v", got, test.Want)
-			}
-		})
-	}
-
 }

--- a/internal/configs/experiments.go
+++ b/internal/configs/experiments.go
@@ -198,7 +198,7 @@ func checkModuleExperiments(m *Module) hcl.Diagnostics {
 
 	if !m.ActiveExperiments.Has(experiments.ModuleVariableOptionalAttrs) {
 		for _, v := range m.Variables {
-			if typeConstraintHasOptionalAttrs(v.Type) {
+			if typeConstraintHasOptionalAttrs(v.ConstraintType) {
 				diags = diags.Append(&hcl.Diagnostic{
 					Severity: hcl.DiagError,
 					Summary:  "Optional object type attributes are experimental",

--- a/internal/configs/module_merge.go
+++ b/internal/configs/module_merge.go
@@ -51,6 +51,7 @@ func (v *Variable) merge(ov *Variable) hcl.Diagnostics {
 	}
 	if ov.Type != cty.NilType {
 		v.Type = ov.Type
+		v.ConstraintType = ov.ConstraintType
 	}
 	if ov.ParsingMode != 0 {
 		v.ParsingMode = ov.ParsingMode
@@ -67,7 +68,7 @@ func (v *Variable) merge(ov *Variable) hcl.Diagnostics {
 	// constraint but the converted value cannot. In practice, this situation
 	// should be rare since most of our conversions are interchangable.
 	if v.Default != cty.NilVal {
-		val, err := convert.Convert(v.Default, v.Type)
+		val, err := convert.Convert(v.Default, v.ConstraintType)
 		if err != nil {
 			// What exactly we'll say in the error message here depends on whether
 			// it was Default or Type that was overridden here.

--- a/internal/configs/module_merge_test.go
+++ b/internal/configs/module_merge_test.go
@@ -25,6 +25,7 @@ func TestModuleOverrideVariable(t *testing.T) {
 			DescriptionSet: true,
 			Default:        cty.StringVal("b_override"),
 			Type:           cty.String,
+			ConstraintType: cty.String,
 			ParsingMode:    VariableParseLiteral,
 			DeclRange: hcl.Range{
 				Filename: "testdata/valid-modules/override-variable/primary.tf",
@@ -46,6 +47,7 @@ func TestModuleOverrideVariable(t *testing.T) {
 			DescriptionSet: true,
 			Default:        cty.StringVal("b_override partial"),
 			Type:           cty.String,
+			ConstraintType: cty.String,
 			ParsingMode:    VariableParseLiteral,
 			DeclRange: hcl.Range{
 				Filename: "testdata/valid-modules/override-variable/primary.tf",

--- a/internal/configs/named_values.go
+++ b/internal/configs/named_values.go
@@ -22,7 +22,13 @@ type Variable struct {
 	Name        string
 	Description string
 	Default     cty.Value
-	Type        cty.Type
+
+	// Type is the concrete type of the variable value.
+	Type cty.Type
+	// ConstraintType is used for decoding and type conversions, and may
+	// contain nested ObjectWithOptionalAttr types.
+	ConstraintType cty.Type
+
 	ParsingMode VariableParsingMode
 	Validations []*VariableValidation
 	Sensitive   bool
@@ -45,6 +51,7 @@ func decodeVariableBlock(block *hcl.Block, override bool) (*Variable, hcl.Diagno
 	// or not they are set when we merge.
 	if !override {
 		v.Type = cty.DynamicPseudoType
+		v.ConstraintType = cty.DynamicPseudoType
 		v.ParsingMode = VariableParseLiteral
 	}
 
@@ -92,7 +99,8 @@ func decodeVariableBlock(block *hcl.Block, override bool) (*Variable, hcl.Diagno
 	if attr, exists := content.Attributes["type"]; exists {
 		ty, parseMode, tyDiags := decodeVariableType(attr.Expr)
 		diags = append(diags, tyDiags...)
-		v.Type = ty
+		v.ConstraintType = ty
+		v.Type = ty.WithoutOptionalAttributesDeep()
 		v.ParsingMode = parseMode
 	}
 
@@ -112,9 +120,9 @@ func decodeVariableBlock(block *hcl.Block, override bool) (*Variable, hcl.Diagno
 		// attribute above.
 		// However, we can't do this if we're in an override file where
 		// the type might not be set; we'll catch that during merge.
-		if v.Type != cty.NilType {
+		if v.ConstraintType != cty.NilType {
 			var err error
-			val, err = convert.Convert(val, v.Type)
+			val, err = convert.Convert(val, v.ConstraintType)
 			if err != nil {
 				diags = append(diags, &hcl.Diagnostic{
 					Severity: hcl.DiagError,

--- a/internal/plans/objchange/objchange_test.go
+++ b/internal/plans/objchange/objchange_test.go
@@ -1452,12 +1452,12 @@ func TestProposedNew(t *testing.T) {
 				"map":    cty.NullVal(cty.Map(cty.Object(map[string]cty.Type{"bar": cty.String}))),
 				"set":    cty.NullVal(cty.Set(cty.Object(map[string]cty.Type{"bar": cty.String}))),
 				"nested_map": cty.NullVal(cty.Map(cty.Object(map[string]cty.Type{
-					"inner": cty.ObjectWithOptionalAttrs(map[string]cty.Type{
+					"inner": cty.Object(map[string]cty.Type{
 						"optional":          cty.String,
 						"computed":          cty.String,
 						"optional_computed": cty.String,
 						"required":          cty.String,
-					}, []string{"computed", "optional", "optional_computed"}),
+					}),
 				}))),
 			}),
 		},

--- a/internal/terraform/evaluate.go
+++ b/internal/terraform/evaluate.go
@@ -238,7 +238,16 @@ func (d *evaluationStateData) GetInputVariable(addr addrs.InputVariable, rng tfd
 		return cty.DynamicVal, diags
 	}
 
+	// wantType is the concrete value type to be returned.
 	wantType := cty.DynamicPseudoType
+
+	// converstionType is the type used for conversion, which may include
+	// optional attributes.
+	conversionType := cty.DynamicPseudoType
+
+	if config.ConstraintType != cty.NilType {
+		conversionType = config.ConstraintType
+	}
 	if config.Type != cty.NilType {
 		wantType = config.Type
 	}
@@ -282,7 +291,7 @@ func (d *evaluationStateData) GetInputVariable(addr addrs.InputVariable, rng tfd
 	}
 
 	var err error
-	val, err = convert.Convert(val, wantType)
+	val, err = convert.Convert(val, conversionType)
 	if err != nil {
 		// We should never get here because this problem should've been caught
 		// during earlier validation, but we'll do something reasonable anyway.

--- a/internal/terraform/evaluate.go
+++ b/internal/terraform/evaluate.go
@@ -237,21 +237,6 @@ func (d *evaluationStateData) GetInputVariable(addr addrs.InputVariable, rng tfd
 		})
 		return cty.DynamicVal, diags
 	}
-
-	// wantType is the concrete value type to be returned.
-	wantType := cty.DynamicPseudoType
-
-	// converstionType is the type used for conversion, which may include
-	// optional attributes.
-	conversionType := cty.DynamicPseudoType
-
-	if config.ConstraintType != cty.NilType {
-		conversionType = config.ConstraintType
-	}
-	if config.Type != cty.NilType {
-		wantType = config.Type
-	}
-
 	d.Evaluator.VariableValuesLock.Lock()
 	defer d.Evaluator.VariableValuesLock.Unlock()
 
@@ -271,15 +256,15 @@ func (d *evaluationStateData) GetInputVariable(addr addrs.InputVariable, rng tfd
 	if d.Operation == walkValidate {
 		// Ensure variable sensitivity is captured in the validate walk
 		if config.Sensitive {
-			return cty.UnknownVal(wantType).Mark(marks.Sensitive), diags
+			return cty.UnknownVal(config.Type).Mark(marks.Sensitive), diags
 		}
-		return cty.UnknownVal(wantType), diags
+		return cty.UnknownVal(config.Type), diags
 	}
 
 	moduleAddrStr := d.ModulePath.String()
 	vals := d.Evaluator.VariableValues[moduleAddrStr]
 	if vals == nil {
-		return cty.UnknownVal(wantType), diags
+		return cty.UnknownVal(config.Type), diags
 	}
 
 	val, isSet := vals[addr.Name]
@@ -287,11 +272,11 @@ func (d *evaluationStateData) GetInputVariable(addr addrs.InputVariable, rng tfd
 		if config.Default != cty.NilVal {
 			return config.Default, diags
 		}
-		return cty.UnknownVal(wantType), diags
+		return cty.UnknownVal(config.Type), diags
 	}
 
 	var err error
-	val, err = convert.Convert(val, conversionType)
+	val, err = convert.Convert(val, config.ConstraintType)
 	if err != nil {
 		// We should never get here because this problem should've been caught
 		// during earlier validation, but we'll do something reasonable anyway.
@@ -303,7 +288,7 @@ func (d *evaluationStateData) GetInputVariable(addr addrs.InputVariable, rng tfd
 		})
 		// Stub out our return value so that the semantic checker doesn't
 		// produce redundant downstream errors.
-		val = cty.UnknownVal(wantType)
+		val = cty.UnknownVal(config.Type)
 	}
 
 	// Mark if sensitive

--- a/internal/terraform/evaluate_test.go
+++ b/internal/terraform/evaluate_test.go
@@ -95,15 +95,19 @@ func TestEvaluatorGetInputVariable(t *testing.T) {
 			Module: &configs.Module{
 				Variables: map[string]*configs.Variable{
 					"some_var": {
-						Name:      "some_var",
-						Sensitive: true,
-						Default:   cty.StringVal("foo"),
+						Name:           "some_var",
+						Sensitive:      true,
+						Default:        cty.StringVal("foo"),
+						Type:           cty.String,
+						ConstraintType: cty.String,
 					},
 					// Avoid double marking a value
 					"some_other_var": {
-						Name:      "some_other_var",
-						Sensitive: true,
-						Default:   cty.StringVal("bar"),
+						Name:           "some_other_var",
+						Sensitive:      true,
+						Default:        cty.StringVal("bar"),
+						Type:           cty.String,
+						ConstraintType: cty.String,
 					},
 				},
 			},

--- a/internal/terraform/node_module_variable.go
+++ b/internal/terraform/node_module_variable.go
@@ -200,7 +200,6 @@ func (n *nodeModuleVariable) DotNode(name string, opts *dag.DotOpts) *dag.DotNod
 // validation, and we will not have any expansion module instance
 // repetition data.
 func (n *nodeModuleVariable) evalModuleCallArgument(ctx EvalContext, validateOnly bool) (map[string]cty.Value, error) {
-	wantType := n.Config.Type
 	name := n.Addr.Variable.Name
 	expr := n.Expr
 
@@ -238,7 +237,7 @@ func (n *nodeModuleVariable) evalModuleCallArgument(ctx EvalContext, validateOnl
 	// now we can do our own local type conversion and produce an error message
 	// with better context if it fails.
 	var convErr error
-	val, convErr = convert.Convert(val, wantType)
+	val, convErr = convert.Convert(val, n.Config.ConstraintType)
 	if convErr != nil {
 		diags = diags.Append(&hcl.Diagnostic{
 			Severity: hcl.DiagError,
@@ -251,7 +250,7 @@ func (n *nodeModuleVariable) evalModuleCallArgument(ctx EvalContext, validateOnl
 		})
 		// We'll return a placeholder unknown value to avoid producing
 		// redundant downstream errors.
-		val = cty.UnknownVal(wantType)
+		val = cty.UnknownVal(n.Config.Type)
 	}
 
 	vals := make(map[string]cty.Value)

--- a/internal/terraform/node_module_variable_test.go
+++ b/internal/terraform/node_module_variable_test.go
@@ -7,6 +7,7 @@ import (
 	"github.com/go-test/deep"
 	"github.com/hashicorp/hcl/v2"
 	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/zclconf/go-cty/cty"
 
 	"github.com/hashicorp/terraform/internal/addrs"
 	"github.com/hashicorp/terraform/internal/configs"
@@ -16,7 +17,9 @@ func TestNodeModuleVariablePath(t *testing.T) {
 	n := &nodeModuleVariable{
 		Addr: addrs.RootModuleInstance.InputVariable("foo"),
 		Config: &configs.Variable{
-			Name: "foo",
+			Name:           "foo",
+			Type:           cty.String,
+			ConstraintType: cty.String,
 		},
 	}
 
@@ -31,7 +34,9 @@ func TestNodeModuleVariableReferenceableName(t *testing.T) {
 	n := &nodeExpandModuleVariable{
 		Addr: addrs.InputVariable{Name: "foo"},
 		Config: &configs.Variable{
-			Name: "foo",
+			Name:           "foo",
+			Type:           cty.String,
+			ConstraintType: cty.String,
 		},
 	}
 
@@ -64,7 +69,9 @@ func TestNodeModuleVariableReference(t *testing.T) {
 		Addr:   addrs.InputVariable{Name: "foo"},
 		Module: addrs.RootModule.Child("bar"),
 		Config: &configs.Variable{
-			Name: "foo",
+			Name:           "foo",
+			Type:           cty.String,
+			ConstraintType: cty.String,
 		},
 		Expr: &hclsyntax.ScopeTraversalExpr{
 			Traversal: hcl.Traversal{
@@ -90,7 +97,9 @@ func TestNodeModuleVariableReference_grandchild(t *testing.T) {
 		Addr:   addrs.InputVariable{Name: "foo"},
 		Module: addrs.RootModule.Child("bar"),
 		Config: &configs.Variable{
-			Name: "foo",
+			Name:           "foo",
+			Type:           cty.String,
+			ConstraintType: cty.String,
 		},
 		Expr: &hclsyntax.ScopeTraversalExpr{
 			Traversal: hcl.Traversal{

--- a/internal/terraform/node_root_variable_test.go
+++ b/internal/terraform/node_root_variable_test.go
@@ -5,6 +5,7 @@ import (
 
 	"github.com/hashicorp/terraform/internal/addrs"
 	"github.com/hashicorp/terraform/internal/configs"
+	"github.com/zclconf/go-cty/cty"
 )
 
 func TestNodeRootVariableExecute(t *testing.T) {
@@ -13,7 +14,9 @@ func TestNodeRootVariableExecute(t *testing.T) {
 	n := &NodeRootVariable{
 		Addr: addrs.InputVariable{Name: "foo"},
 		Config: &configs.Variable{
-			Name: "foo",
+			Name:           "foo",
+			Type:           cty.String,
+			ConstraintType: cty.String,
 		},
 	}
 

--- a/internal/terraform/variables.go
+++ b/internal/terraform/variables.go
@@ -262,10 +262,8 @@ func checkInputVariables(vcs map[string]*configs.Variable, vs InputValues) tfdia
 			continue
 		}
 
-		wantType := vc.Type
-
 		// A given value is valid if it can convert to the desired type.
-		_, err := convert.Convert(val.Value, wantType)
+		_, err := convert.Convert(val.Value, vc.ConstraintType)
 		if err != nil {
 			switch val.SourceType {
 			case ValueFromConfig, ValueFromAutoFile, ValueFromNamedFile:

